### PR TITLE
refactor rafthttp

### PR DESCRIPTION
**Pull request by xiang90**

### CODE REVIEW

File: rafthttp/streamer.go [hunk 1]
```diff
@@ -41,60 +42,69 @@ const (
...
+	stopped bool
 }
 
 func (s *stream) open(from, to, cid types.ID, term uint64, tr http.RoundTripper, u string, r Raft) error {
-	if s.client != nil {
-		panic("open: stream is open")
-	}
-
-	c, err := newStreamClient(from, to, cid, term, tr, u, r)
+	c, err := newStreamReader(from, to, cid, term, tr, u, r)
```
Comment (yichengq) [inline rafthttp/streamer.go]: s/c/r

File: rafthttp/streamer.go [hunk 2]
```diff
@@ -41,60 +42,69 @@ const (
...
 	if err != nil {
 		log.Printf("stream: error opening stream: %v", err)
 		return err
 	}
-	s.client = c
+
+	s.Lock()
+	defer s.Unlock()
+	if s.stopped {
```
Comment (yichengq) [inline rafthttp/streamer.go]: if we check stopped here, we should close the inited stream reader.

**Comment by yichengq** [comment]
squash and lgmt

**Comment by xiang90** [comment]
@yichengq There is no need for squash. Each commit does one thing.

**Pull request merged**